devito/finite_differences/differentiable.py

```diff
@@ -423,14 +423,30 @@ def __new__(cls, *args, **kwargs):
         # a*1 -> a
         args = [i for i in args if i != 1]
 
+        # a*-1 -> a
         # a*-1*-1 -> a
+        # a*-1*-1*-1 -> a*-1
         nminus = len([i for i in args if i == sympy.S.NegativeOne])
-        if nminus % 2 == 0:
-            args = [i for i in args if i != sympy.S.NegativeOne]
+        args = [i for i in args if i != sympy.S.NegativeOne]
+        if nminus % 2 == 1:
+            args.append(sympy.S.NegativeOne)
 
         # Reorder for homogeneity with pure SymPy types
         _mulsort(args)
 
+        # `sympy.Mul.flatten(coeff, Add)` flattens out nested Adds within Add,
+        # which would destroy `EvalDerivative`s if present. So here we perform
+        # a similar thing, but cautiously construct an evaluated Add, which
+        # will preserve the integrity of `EvalDerivative`s, if any.
+        try:
+            a, b = args
+            if not a.is_zero and a.is_Rational:
+                r, b = b.as_coeff_Mul()
+                if r is sympy.S.One and type(b) is Add:
```

Comment:
Can this be generalized to like `r is sympy.Number` or `r.is_Symbol` and merge it with `a`?

Reply:
for the record, I lifted it from sympy

https://github.com/sympy/sympy/blob/master/sympy/core/mul.py#L285

tho admittedly a slightly older version than the one I have installed locally

I will try to improve it